### PR TITLE
Http mapper tests

### PR DESCRIPTION
**Pull request by denysov-serhii**

### CODE REVIEW

File: tests/ArraySerializableHelperTrait.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+use Infrastructure\Models\ArraySerializable;
+use Infrastructure\Models\Collection;
+
+trait ArraySerializableHelperTrait
+{
+    /**
+     * @param array $data
+     * @return Collection
+     */
+    private function createArraySerializableCollection(array $data): Collection
```
Comment (phpadventure) [inline tests/ArraySerializableHelperTrait.php:15]: please don't use name all inheritance of class.

File: tests/Mappers/BaseMapperTest.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ * Test for BaseMapper class
+ * @see BaseMapper
+ */
+class BaseMapperTest extends TestCase
+{
+    use ReflectionHelperTrait;
+    use ArraySerializableHelperTrait;
+
+    /**
+     * @see BaseMapper::buildPaginationCollection()
```
Comment (phpadventure) [inline tests/Mappers/BaseMapperTest.php:22]: please synchronize what are you testing

File: tests/Models/Http/HeadersTest.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+     */
+    public function testHasHeader()
+    {
+        $data = [
+            'X-TEST-1' => ['val_1', 'val_2'],
+            'X-TEST-2' => ['val_3', 'val_4'],
+            'X-TEST-3' => ['val_5'],
+        ];
+        /** @var Headers $headers */
+        $headers = $this->createInstance(Headers::class);
```
Comment (phpadventure) [inline tests/Models/Http/HeadersTest.php:31]: What is the reason create to  object so weird ? 

What is main purpose of this test? 

What is diff if you create object in natural way? 

If __construct doesn't work your test will fail which is correct. 

This code is redundant and hiding normal behavior.

File: tests/Models/Http/HeadersTest.php [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        $originalHeaders1 = new Headers(['H1' => 'v1', 'H2' => 'v2', 'COMMON' => 'headers_1', 'DIFFERENT_1' => '1']);
+        $originalHeaders2 = new Headers(['H3' => 'v3', 'H4' => 'v4', 'COMMON' => 'headers_2', 'DIFFERENT_2' => '2']);
+
+        $headers1 = clone $originalHeaders1;
+        $headers2 = clone $originalHeaders2;
+
+        $mergedHeaders = $headers1->merge($headers2);
+
+        $this->assertEquals($originalHeaders1, $headers1);
+        $this->assertEquals($originalHeaders2, $headers2);
```
Comment (phpadventure) [inline tests/Models/Http/HeadersTest.php:58]: what is reason to test clone ?

File: tests/Models/Http/HeadersTest.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        $this->assertEquals(
+            $this->getProperty($mergedHeaders, 'headersContainer')['COMMON'],
+            $this->getProperty($headers2, 'headersContainer')['COMMON']
+        );
+
+        $this->assertTrue(
+            array_key_exists('DIFFERENT_1', $this->getProperty($mergedHeaders, 'headersContainer'))
+        );
+        $this->assertTrue(
+            array_key_exists('DIFFERENT_2', $this->getProperty($mergedHeaders, 'headersContainer'))
```
Comment (phpadventure) [inline tests/Models/Http/HeadersTest.php:69]: The purpose of this test is to check if all items merged. You check if only two elements mergred.

File: tests/Models/Http/HttpClientTest.php [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+ * @see HttpClient
+ */
+class HttpClientTest extends TestCase
+{
+    use ReflectionHelperTrait;
+
+    /**
+     * @throws \ReflectionException
+     */
+    public function test__construct()
```
Comment (phpadventure) [inline tests/Models/Http/HttpClientTest.php:30]: what is the purpose of this test?

File: tests/ReflectionHelperTrait.php [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+trait ReflectionHelperTrait
+{
+    /**
+     * @param $object
+     * @param string $method
+     * @param mixed ...$arguments
+     * @return mixed
+     * @throws \ReflectionException
+     */
+    private function callMethod($object, string $method, ...$arguments)
```
Comment (phpadventure) [inline tests/ReflectionHelperTrait.php:15]: Please get rid of reflections

File: tests/ReflectionHelperTrait.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
+<?php
+
+namespace Infrastructure\Tests;
+
+
+trait ReflectionHelperTrait
```
Comment (phpadventure) [inline tests/ReflectionHelperTrait.php:6]: please get rid of traits

File: tests/ReflectionHelperTrait.php [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+        return $object;
+    }
+
+    /**
+     * @param string $class
+     * @return object
+     * @throws \ReflectionException
+     */
+    private function createInstance(string $class)
```
Comment (phpadventure) [inline tests/ReflectionHelperTrait.php:116]: please use fabric method in test, DI. but please get rid of this.

File: tests/Models/Http/HttpClientTest.php (hunk 6 above)
Comment (phpadventure) [reply tests/Models/Http/HttpClientTest.php:30]: The main purpose of test is to cover most critical in core behavior of class. You will never cover all cases. You need to find balance. When you are writing test please figure what are the purpose of test. 
If you hack your classes it means you need to make refactor.

File: Mappers/HttpMapper.php [hunk 10]
```diff
@@ -214,12 +214,12 @@ private function prepareParams(SearchCriteria $filter): array
      * @return RequestInterface
      * @throws \Infrastructure\Models\Http\IllegalHeaderValueException
      */
-    private function mergeDefaultData(RequestInterface $request)
+    protected function mergeDefaultData(RequestInterface $request)
```
Comment (Antonyan) [inline Mappers/HttpMapper.php:217]: pls, add return types

File: tests/ArraySerializableHelperTrait.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+namespace Infrastructure\Tests;
+
+
+use Infrastructure\Models\ArraySerializable;
+use Infrastructure\Models\Collection;
+
+trait ArraySerializableHelperTrait
```
Comment (Antonyan) [inline tests/ArraySerializableHelperTrait.php:9]: ?????

File: tests/Mappers/BaseMapperTest.php [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    public function testBuildCollection()
+    {
+        $baseMapper = $this->getMockBuilder(BaseMapper::class)
```
Comment (Antonyan) [inline tests/Mappers/BaseMapperTest.php:26]: Why are you testing abstract method?

File: tests/Mappers/BaseMapperTest.php [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        $result = $this->callMethod($baseMapper, 'buildCollection', $testData);
+
+        $this->assertInstanceOf(Collection::class, $result);
+        $this->assertEquals($this->createArraySerializableCollection($testData), $result);
+    }
+
+    /**
+     * @see BaseMapper::buildPaginationCollection()
+     */
+    public function testBuildPaginationCollection()
```
Comment (Antonyan) [inline tests/Mappers/BaseMapperTest.php:52]: protected method shouldn't be tested

**Pull request closed without merging**